app/javascript/components/common/searchable-list/FilterPanel.tsx

```diff
@@ -0,0 +1,109 @@
+import React, { useState, useEffect, useRef, useCallback } from 'react'
+import { FilterList } from './FilterList'
+import { GraphicalIcon } from '../GraphicalIcon'
+import { FilterCategory, FilterValue } from '../SearchableList'
+
+export const FilterPanel = ({
+  value = {},
+  categories,
+  setFilter,
+}: {
+  value?: FilterValue
+  categories: FilterCategory[]
+  setFilter: (value: FilterValue) => void
+}): JSX.Element => {
+  const [expanded, setExpanded] = useState(false)
+  const [selected, setSelected] = useState(value)
+  const [hasExpandedEver, markAsExpanded] = useState(false)
+
+  const dialogRef = useRef<HTMLDivElement>(null)
+  const filterButtonRef = useRef<HTMLButtonElement>(null)
+
+  useEffect(() => {
+    if (!expanded || !dialogRef.current) {
+      return
+    }
+
+    dialogRef.current.focus()
+    markAsExpanded(true)
+  }, [expanded])
+
+  useEffect(() => {
+    if (!hasExpandedEver || !filterButtonRef.current) {
+      return
+    }
+    filterButtonRef.current.focus()
+  }, [hasExpandedEver])
+
+  useEffect(() => {
+    if (!expanded) {
+      return
+    }
+
+    const handleEscape = (e: KeyboardEvent) => {
+      if (e.key !== 'Escape') {
+        return
+      }
+
+      e.stopPropagation()
+      setExpanded(false)
+    }
+
+    document.addEventListener('keyup', handleEscape)
+    return () => {
+      document.removeEventListener('keyup', handleEscape)
+    }
+  }, [expanded])
+
+  const handleSubmit = useCallback(() => {
+    setFilter(selected)
+    setExpanded(false)
+  }, [selected, setFilter])
+
+  const handleClose = useCallback(() => {
+    setExpanded(false)
+  }, [])
+
+  return (
+    <>
+      <button
+        ref={filterButtonRef}
+        onClick={() => setExpanded(true)}
+        className="--filter-btn"
+        aria-haspopup="true"
+        aria-expanded={expanded}
+      >
+        Filter by
+        <GraphicalIcon icon="chevron-down" />
+      </button>
+      <div
+        ref={dialogRef}
+        tabIndex={-1}
+        role="dialog"
+        aria-label="A series of checkboxes to filter solutions"
```

Comment:
Is this aria label still correct, or should it be parameterized now that this looks like a general component?

Reply:
I will change this :) Good catch.